// xclip/dist/cjs/os.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.runCommand = exports.getShell = exports.getCurrentPlatform = void 0;
const os = __importStar(require("os"));
const child_process_1 = require("child_process");
const is_wsl_1 = __importDefault(require("is-wsl"));
const win10_1 = require("./clipboard/win10");
const win32_1 = require("./clipboard/win32");
const wsl_1 = require("./clipboard/wsl");
const linux_1 = require("./clipboard/linux");
const darwin_1 = require("./clipboard/darwin");
function getCurrentPlatform() {
    const platform = process.platform;
    if (is_wsl_1.default) {
        return "wsl";
    }
    if (platform === "win32") {
        const currentOS = os.release().split(".")[0];
        if (currentOS === "10") {
            return "win10";
        }
        else {
            return "win32";
        }
    }
    else if (platform === "darwin") {
        return "darwin";
    }
    else {
        return "linux";
    }
}
exports.getCurrentPlatform = getCurrentPlatform;
function getShell() {
    const platform = getCurrentPlatform();
    switch (platform) {
        case "win10":
            return new Win10Shell();
        case "linux":
            return new LinuxShell();
        case "darwin":
            return new DarwinShell();
        case "win32":
            return new Win32Shell();
        case "wsl":
            return new WslShell();
        default:
            throw new Error("Unsupported platform");
    }
}
exports.getShell = getShell;
function runCommand(shell, options, timeout = 10000) {
    return new Promise((resolve, reject) => {
        let errorTriggered = false;
        let output = "";
        let errorMessage = "";
        const process = (0, child_process_1.spawn)(shell, options, { timeout });
        process.stdout.on("data", (chunk) => {
            output += `${chunk}`;
        });
        process.stderr.on("data", (chunk) => {
            errorMessage += `${chunk}`;
        });
        process.on("exit", (code) => {
            if (process.killed) {
                console.log("Process took too long and was killed");
            }
            if (!errorTriggered) {
                if (code === 0) {
                    resolve(output);
                }
                else {
                    reject(errorMessage);
                }
            }
        });
        process.on("error", (error) => {
            errorTriggered = true;
            reject(error);
        });
    });
}
exports.runCommand = runCommand;
class Win10Shell {
    getClipboard() {
        return new win10_1.Win10Clipboard();
    }
    async runScript(script, parameters) {
        const shell = "powershell";
        const command = [
            "-noprofile",
            "-noninteractive",
            "-nologo",
            "-sta",
            "-executionpolicy",
            "bypass",
            "-windowstyle",
            "hidden",
            "-file",
            script,
        ].concat(parameters);
        const stdout = await runCommand(shell, command);
        return stdout;
    }
}
class Win32Shell {
    getClipboard() {
        return new win32_1.Win32Clipboard();
    }
    async runScript(script, parameters) {
        const shell = "powershell";
        const command = [
            "-noprofile",
            "-noninteractive",
            "-nologo",
            "-sta",
            "-executionpolicy",
            "bypass",
            "-windowstyle",
            "hidden",
            "-file",
            script,
        ].concat(parameters);
        const stdout = await runCommand(shell, command);
        return stdout;
    }
}
class WslShell {
    getClipboard() {
        return new wsl_1.WslClipboard();
    }
    async runScript(script, parameters) {
        const shell = "/mnt/c/Windows/System32/WindowsPowerShell/v1.0/powershell.exe";
        const command = [
            "-noprofile",
            "-noninteractive",
            "-nologo",
            "-sta",
            "-executionpolicy",
            "bypass",
            "-windowstyle",
            "hidden",
            "-file",
            script,
        ].concat(parameters);
        const stdout = await runCommand(shell, command);
        return stdout;
    }
}
class LinuxShell {
    getClipboard() {
        return new linux_1.LinuxClipboard();
    }
    async runScript(script, parameters) {
        const shell = "sh";
        const command = [script].concat(parameters);
        const stdout = await runCommand(shell, command);
        return stdout;
    }
}
class DarwinShell {
    getClipboard() {
        return new darwin_1.DarwinClipboard();
    }
    async runScript(script, parameters) {
        const shell = "osascript";
        const command = [script].concat(parameters);
        const stdout = await runCommand(shell, command);
        return stdout;
    }
}
