// xclip/dist/esm/clipboard/darwin.js
import { fileURLToPath } from "node:url";
import { ClipboardType } from "../clipboard_interface";
import { getShell } from "../os";
import * as path from "path";
import { stripFinalNewline } from "../utils";
import { BaseClipboard } from "./base_clipboard";
function darwin_HextoHtml(str) {
    const regex = /«data HTML(.*?)»/;
    const subst = `$1`;
    const data = str.replace(regex, subst);
    const buff = Buffer.from(data, "hex");
    return buff.toString("utf8");
}
class DarwinClipboard extends BaseClipboard {
    SCRIPT_PATH = "../../res/scripts/";
    async copyImage(imageFile) {
        const imageFilePath = fileURLToPath(imageFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_set_clipboard_png.applescript");
        const params = [imageFilePath];
        try {
            const shell = getShell();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async copyTextPlain(textFile) {
        const textFilePath = fileURLToPath(textFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_set_clipboard_text_plain.applescript");
        const params = [textFilePath];
        try {
            const shell = getShell();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async copyTextHtml(htmlFile) {
        const htmlFilePath = fileURLToPath(htmlFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_set_clipboard_text_html.applescript");
        const params = [htmlFilePath];
        try {
            const shell = getShell();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    onDetectType(types) {
        const detectedTypes = new Set();
        for (const type of types) {
            switch (type) {
                case "Text":
                    detectedTypes.add(ClipboardType.Text);
                    break;
                case "HTML":
                    detectedTypes.add(ClipboardType.Html);
                    break;
                case "Image":
                    detectedTypes.add(ClipboardType.Image);
                    break;
            }
        }
        return detectedTypes;
    }
    async getContentType() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_get_clipboard_content_type.applescript");
        try {
            const shell = getShell();
            const data = await shell.runScript(script);
            console.debug("getClipboardContentType", data);
            const types = data.split(/\r\n|\n|\r/);
            return this.detectType(types);
        }
        catch (e) {
            return ClipboardType.Unknown;
        }
    }
    async getImage(imagePath) {
        if (!imagePath)
            return "";
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_save_clipboard_png.applescript");
        const shell = getShell();
        const data = await shell.runScript(script, [imagePath]);
        return stripFinalNewline(data);
    }
    async getTextPlain() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_get_clipboard_text_plain.applescript");
        const shell = getShell();
        const data = await shell.runScript(script);
        return stripFinalNewline(data);
    }
    async getTextHtml() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "darwin_get_clipboard_text_html.applescript");
        const shell = getShell();
        const data = await shell.runScript(script);
        return darwin_HextoHtml(stripFinalNewline(data));
    }
}
export { DarwinClipboard };
