// xclip/dist/cjs/clipboard/linux.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinuxClipboard = void 0;
const node_url_1 = require("node:url");
const clipboard_interface_1 = require("../clipboard_interface");
const os_1 = require("../os");
const path = __importStar(require("path"));
const utils_1 = require("../utils");
const base_clipboard_1 = require("./base_clipboard");
class LinuxClipboard extends base_clipboard_1.BaseClipboard {
    SCRIPT_PATH = "../../res/scripts/";
    async copyImage(imageFile) {
        const imageFilePath = (0, node_url_1.fileURLToPath)(imageFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_set_clipboard_png.sh");
        const params = [imageFilePath];
        try {
            const shell = (0, os_1.getShell)();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async copyTextPlain(textFile) {
        const textFilePath = (0, node_url_1.fileURLToPath)(textFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_set_clipboard_text_plain.sh");
        const params = [textFilePath];
        try {
            const shell = (0, os_1.getShell)();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async copyTextHtml(htmlFile) {
        const htmlFilePath = (0, node_url_1.fileURLToPath)(htmlFile);
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_set_clipboard_text_html.sh");
        const params = [htmlFilePath];
        try {
            const shell = (0, os_1.getShell)();
            await shell.runScript(script, params);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    onDetectType(types) {
        const detectedTypes = new Set();
        for (const type of types) {
            switch (type) {
                case "no xclip":
                    console.error("You need to install xclip command first.");
                    detectedTypes.add(clipboard_interface_1.ClipboardType.Unknown);
                    return detectedTypes;
                case "image/png":
                    detectedTypes.add(clipboard_interface_1.ClipboardType.Image);
                    break;
                case "text/html":
                    detectedTypes.add(clipboard_interface_1.ClipboardType.Html);
                    break;
                default:
                    detectedTypes.add(clipboard_interface_1.ClipboardType.Text);
                    break;
            }
        }
        return detectedTypes;
    }
    async getContentType() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_get_clipboard_content_type.sh");
        try {
            const shell = (0, os_1.getShell)();
            const data = await shell.runScript(script);
            console.debug("getClipboardContentType", data);
            const types = data.split(/\r\n|\n|\r/);
            return this.detectType(types);
        }
        catch (e) {
            return clipboard_interface_1.ClipboardType.Unknown;
        }
    }
    async getImage(imagePath) {
        if (!imagePath)
            return "";
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_save_clipboard_png.sh");
        const shell = (0, os_1.getShell)();
        const data = await shell.runScript(script, [imagePath]);
        return (0, utils_1.stripFinalNewline)(data);
    }
    async getTextPlain() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_get_clipboard_text_plain.sh");
        const shell = (0, os_1.getShell)();
        const data = await shell.runScript(script);
        return (0, utils_1.stripFinalNewline)(data);
    }
    async getTextHtml() {
        const script = path.join(__dirname, this.SCRIPT_PATH, "linux_get_clipboard_text_html.sh");
        const shell = (0, os_1.getShell)();
        const data = await shell.runScript(script);
        return (0, utils_1.stripFinalNewline)(data);
    }
}
exports.LinuxClipboard = LinuxClipboard;
